README.md
# NAME

Mojo::Netdata - https://netdata.cloud plugin for Perl

# SYNOPSIS

## Installation

    sudo -i
    apt install -y cpanminus
    cpanm -n https://github.com/jhthorsen/mojo-netdata/archive/refs/heads/main.tar.gz
    ln -s $(which mojo-netdata) /etc/netdata/custom-plugins.d/mojo-netdata.plugin

    # See "Config file" below for information on what to place inside mojo.conf.pl
    $EDITOR /etc/netdata/mojo.conf.pl

## Config files

The config files are located in `/etc/netdata/mojo.conf.d`. The files are
plain Perl files, which means you can define variables and call functions. The
only important part is that the last statement in the file is a hash-ref.

Any hash-refs that has the "collector" key will be placed into ["collectors"](#collectors),
while everything else will be merged into ["config"](#config). Example:

    # /etc/netdata/mojo.conf.d/foo.conf.pl
    {foo => 42, bar => 100}

    # /etc/netdata/mojo.conf.d/bar.conf.pl
    {collector => 'Mojo::Netdata::Collector::HTTP', jobs => []}

The two config files above will result in this ["config"](#config):

    {
      foo => 42,
      bar => 100,
      collectors => [
        {collector => 'Mojo::Netdata::Collector::HTTP', jobs => []},
      },
    }

See ["SYNOPSIS" in Mojo::Netdata::Collector::HTTP](https://metacpan.org/pod/Mojo%3A%3ANetdata%3A%3ACollector%3A%3AHTTP#SYNOPSIS) for an example config file.

## Log file

The output from this Netdata plugin can be found in
`/var/log/netdata/error.log`.

# DESCRIPTION

[Mojo::Netdata](https://metacpan.org/pod/Mojo%3A%3ANetdata) is a plugin for [Netdata](https://netdata.cloud). It can load
custom [Mojo::Netdata::Collector](https://metacpan.org/pod/Mojo%3A%3ANetdata%3A%3ACollector) classes and write data back to Netdata on a
given interval.

This module is currently EXPERIMENTAL, and the API might change without
warning.

# ATTRIBUTES

## cache\_dir

    $path = $netdata->cache_dir;

Holds the `NETDATA_CACHE_DIR` environment variable. See
[https://learn.netdata.cloud/docs/agent/collectors/plugins.d#environment-variables](https://learn.netdata.cloud/docs/agent/collectors/plugins.d#environment-variables)
for more details.

## config

    $hash_ref = $netdata->config;

Holds the config for [Mojo::Netdata](https://metacpan.org/pod/Mojo%3A%3ANetdata) and all ["collectors"](#collectors).

## collectors

    $array_ref = $netdata->collectors;

An array-ref of [Mojo::Netdata::Collector](https://metacpan.org/pod/Mojo%3A%3ANetdata%3A%3ACollector) objects.

## debug\_flags

    $str = $netdata->debug_flags;

Defaults to the `NETDATA_DEBUG_FLAGS` environment variable. See
[https://learn.netdata.cloud/docs/agent/collectors/plugins.d#environment-variables](https://learn.netdata.cloud/docs/agent/collectors/plugins.d#environment-variables)
for more details.

## host\_prefix

    $str = $netdata->host_prefix;

Defaults to the `NETDATA_HOST_PREFIX` environment variable. See
[https://learn.netdata.cloud/docs/agent/collectors/plugins.d#environment-variables](https://learn.netdata.cloud/docs/agent/collectors/plugins.d#environment-variables)
for more details.

## log\_dir

    $path = $netdata->log_dir;

Holds the `NETDATA_LOG_DIR` environment variable. See
[https://learn.netdata.cloud/docs/agent/collectors/plugins.d#environment-variables](https://learn.netdata.cloud/docs/agent/collectors/plugins.d#environment-variables)
for more details.

## plugins\_dir

    $path = $netdata->plugins_dir;

Holds the `NETDATA_PLUGINS_DIR` environment variable. See
[https://learn.netdata.cloud/docs/agent/collectors/plugins.d#environment-variables](https://learn.netdata.cloud/docs/agent/collectors/plugins.d#environment-variables)
for more details.

## stock\_config\_dir

    $path = $netdata->stock_config_dir;

Holds the `NETDATA_STOCK_CONFIG_DIR` environment variable. See
[https://learn.netdata.cloud/docs/agent/collectors/plugins.d#environment-variables](https://learn.netdata.cloud/docs/agent/collectors/plugins.d#environment-variables)
for more details.

## update\_every

    $num = $netdata->update_every;

Defaults to the `NETDATA_UPDATE_EVERY` environment variable. See
[https://learn.netdata.cloud/docs/agent/collectors/plugins.d#environment-variables](https://learn.netdata.cloud/docs/agent/collectors/plugins.d#environment-variables)
for more details.

## user\_config\_dir

    $path = $netdata->user_config_dir;

Holds the `NETDATA_USER_CONFIG_DIR` environment variable. See
[https://learn.netdata.cloud/docs/agent/collectors/plugins.d#environment-variables](https://learn.netdata.cloud/docs/agent/collectors/plugins.d#environment-variables)
for more details.

## web\_dir

    $path = $netdata->web_dir;

Holds the `NETDATA_WEB_DIR` environment variable. See
[https://learn.netdata.cloud/docs/agent/collectors/plugins.d#environment-variables](https://learn.netdata.cloud/docs/agent/collectors/plugins.d#environment-variables)
for more details.

# METHODS

## start

    $bool = $netdata->start;

Reads the ["config"](#config) and return 1 if any ["collectors"](#collectors) got registered.

# AUTHOR

Jan Henning Thorsen

# COPYRIGHT AND LICENSE

Copyright (C) Jan Henning Thorsen

This program is free software, you can redistribute it and/or modify it under
the terms of the Artistic License version 2.0.

# SEE ALSO

[https://learn.netdata.cloud/docs/agent/collectors/plugins.d](https://learn.netdata.cloud/docs/agent/collectors/plugins.d).


Changes
Revision history for perl distribution Mojo-Netdata

0.04 Not Released
 - Add support for NETDATA_DEBUG=1 to avoid logging too much
 - Add support for "concurrency" in Collector::HTTP

0.03 2022-01-31T11:34:51+0900
 - Add support for custom "user_agent" name
 - Add support for "via" in the HTTP collector config
 - Logging more details on failed HTTP request
 - Changed logf() to serialize references

0.02 2022-01-30T17:56:13+0900
 - Add support for reading config from /etc/netdata/mojo.conf.d/
 - Renamed "class" key with "collector"
 - Rewrote the HTTP collector
   * Add support for custom "dimension" name
   * Changed "jobs" into an array-ref
   * Removed support for "direct_ip"
   * The "Host" header can be used as default "family"
 - Improved default context() for Mojo::Netdata::Chart
   to make it easier to define alarms using "template"
 - Removed context() from Mojo::Netdata::Collector

0.01 2022-01-27T11:30:42+0900
 - Add Mojo::Netdata
 - Add Mojo::Netdata::Chart
 - Add Mojo::Netdata::Collector
 - Add Mojo::Netdata::Collector::HTTP


Makefile.PL
# Generated by git-ship. See 'git-ship --man' for help or https://github.com/jhthorsen/app-git-ship
use utf8;
use ExtUtils::MakeMaker;
my %WriteMakefileArgs = (
  NAME           => 'Mojo::Netdata',
  AUTHOR         => 'Jan Henning Thorsen <jhthorsen@cpan.org>',
  LICENSE        => 'artistic_2',
  ABSTRACT_FROM  => 'lib/Mojo/Netdata.pm',
  VERSION_FROM   => 'lib/Mojo/Netdata.pm',
  EXE_FILES      => [qw(script/mojo-netdata)],
  OBJECT         => '',
  BUILD_REQUIRES => {},
  TEST_REQUIRES  => {'Test2::V0'   => '0.000144'},
  PREREQ_PM      => {'Mojolicious' => '8.71', 'perl' => 'v5.22.0'},
  META_MERGE     => {
    'dynamic_config' => 0,
    'meta-spec'      => {version => 2},
    'resources'      => {
      bugtracker => {web => 'https://github.com/jhthorsen/mojo-netdata/issues'},
      homepage   => 'https://github.com/jhthorsen/mojo-netdata',
      repository => {
        type => 'git',
        url  => 'https://github.com/jhthorsen/mojo-netdata.git',
        web  => 'https://github.com/jhthorsen/mojo-netdata',
      },
    },
    'x_contributors' => ['Jan Henning Thorsen'],
  },
  test => {TESTS => (-e 'META.yml' ? 't/*.t' : 't/*.t xt/*.t')},
);

unless (eval { ExtUtils::MakeMaker->VERSION('6.63_03') }) {
  my $test_requires = delete $WriteMakefileArgs{TEST_REQUIRES};
  @{$WriteMakefileArgs{PREREQ_PM}}{keys %$test_requires} = values %$test_requires;
}

WriteMakefile(%WriteMakefileArgs);


lib/Mojo/Netdata/Util.pm
package Mojo::Netdata::Util;
use Mojo::Base -strict, -signatures;

use Exporter qw(import);
use Mojo::File;

our @EXPORT_OK = qw(logf safe_id);

sub logf ($level, $format, @args) {
  return 1 if $ENV{HARNESS_ACTIVE} and !$ENV{HARNESS_IS_VERBOSE};
  return 1 if $level eq 'debug' and !$ENV{HARNESS_IS_VERBOSE} and !$ENV{NETDATA_DEBUG_FLAGS};

  my $module_name = caller;
  my ($s, $m, $h, $day, $month, $year) = localtime time;

  state $program_name = Mojo::File->new($0)->basename;
  printf STDERR "%s-%02s-%02s %02s:%02s:%02s: %s: %s: %s: $format\n", $year + 1900, $month + 1,
    $day, $h, $m, $s, $program_name, uc $level, $module_name, @args;
  return 1;
}

sub safe_id ($str) {
  $str =~ s![^A-Za-z0-9]!_!g;
  $str =~ s!_+$!!g;
  $str =~ s!^_+!!g;
  return $str;
}

1;

=encoding utf8

=head1 NAME

Mojo::Netdata::Util - Utility functions for Mojo::Netdata

=head1 SYNOPSIS

  use Mojo::Netdata::Util qw(safe_id);
  print safe_id 'Not%co.ol';

=head1 DESCRIPTION

L<Mojo::Netdata::Util> as functions that can be useful when working with
L<Mojo::Netdata::Collector> classes.

=head1 EXPORTED FUNCTIONS

=head2 logf

  logf $level, $format, @args;

Used to log messages to STDERR. C<$level> can be "debug", "info", "warnings",
"error", "fatal".

=head2 safe_id

  $str = safe_id $str;

Turns an "unsafe" string into a string you can use for things like "id" or
"type". This is called by L<Mojo::Netdata::Chart/to_string> and
L<Mojo::Netdata::Chart/data_to_string> to make sure the output strings are
safe.

=head1 SEE ALSO

L<Mojo::Netdata>.

=cut


lib/Mojo/Netdata/Chart.pm
package Mojo::Netdata::Chart;
use Mojo::Base -base, -signatures;

use Carp qw(croak);
use Mojo::Netdata::Util qw(safe_id);

has chart_type   => 'line';
has context      => 'default';
has dimensions   => sub ($self) { +{} };
has family       => sub ($self) { $self->id };
has id           => sub ($self) { croak '"id" cannot be built' };
has module       => '';
has name         => '';
has options      => '';                                               # "detail hidden obsolete"
has plugin       => 'mojo';
has priority     => 1000;
has title        => sub ($self) { $self->name || $self->id };
has type         => sub ($self) { croak '"type" cannot be built' };
has units        => '#';
has update_every => 1;

sub data_to_string ($self, $microseconds = undef) {
  my $dimensions = $self->dimensions;
  my $set        = join "\n",
    map { sprintf "SET %s = %s", $_, $dimensions->{$_}{value} // '' } sort keys %$dimensions;

  return !$set ? '' : sprintf "BEGIN %s.%s%s\n%s\nEND\n", safe_id($self->type), safe_id($self->id),
    ($microseconds ? " $microseconds" : ""), $set;
}

sub dimension ($self, $name, $attrs = undef) {
  my $id = safe_id $name;
  return $self->dimensions->{$id} unless $attrs;
  my $dimension = $self->dimensions->{$id} //= {name => $name};
  @$dimension{keys(%$attrs)} = values %$attrs;
  return $self;
}

sub to_string ($self) {
  my $dimensions = $self->dimensions;
  return '' unless %$dimensions;

  my $str = sprintf "CHART %s.%s %s\n", safe_id($self->type), safe_id($self->id),
    q('name' 'title' 'units' 'family' context chart_type priority update_every 'options' 'plugin' 'module')
    =~ s!([a-z_]+)!{$self->$1}!ger;

  for my $id (sort keys %$dimensions) {
    my $dimension = $dimensions->{$id};
    $dimension->{algorithm}  ||= 'absolute';
    $dimension->{divisor}    ||= 1;
    $dimension->{multiplier} ||= 1;
    $dimension->{name}       ||= $id;
    $dimension->{options}    ||= '';
    $str .= sprintf "DIMENSION %s %s\n", $id,
      q('name' algorithm multiplier divisor 'options') =~ s!([a-z_]+)!{$dimension->{$1}}!ger;
  }

  return $str;
}

1;

=encoding utf8

=head1 NAME

Mojo::Netdata::Chart - Represents a Netdata chart and dimensions

=head1 SYNOPSIS

  my $chart = Mojo::Netdata::Chart->new;
  $chart->data_to_string;
  $chart->to_string;

=head1 DESCRIPTION

L<Mojo::Netdata::Chart> is a class that represents a Netdata chart and
dimensions. See L<https://learn.netdata.cloud/docs/agent/collectors/plugins.d#chart>
for more details.

=head1 ATTRIBUTES

=head2 chart_type

  $str = $chart->chart_type;

Either "area", "line" or "stacked". Defaults to "line".

=head2 context

  $str = $chart->context;

Defaults to "default".

=head2 dimensions

  $hash_ref = $chart->dimensions;

See L</dimension>.

=head2 family

  $str = $chart->family;

Defaults to L</id>.

=head2 id

  $str = $chart->id;

Required to be set.

=head2 module

  $str = $chart->module;

Defaults to empty string.

=head2 name

  $str = $chart->name;

Defaults to empty string.

=head2 options

  $str = $chart->options;

Defaults to empty string.

=head2 plugin

  $str = $chart->options;

Defaults to "mojo". The default is subject to change!

=head2 priority

  $int = $chart->priority;

Defaults to 1000.

=head2 title

  $str = $chart->title;

Defaults to L</name> or L</id>.

=head2 type

  $str = $chart->type;

Required to be set.

=head2 units

  $str = $chart->units;

Defaults to "#".

=head2 update_every

  $num = $chart->update_every;

How often to update Netdata.

=head1 METHODS

=head2 data_to_string

  $str = $chart->data_to_string;

Takes the values in L</dimensions> and creates a string with SET, suitable to
be sent to Netdata.

=head2 dimension

  $dimension = $chart->dimension($id);
  $chart = $chart->dimension($id => {name => 'cool'});
  $chart = $chart->dimension($id => {value => 42});

Used to get or set an item in L</dimensions>. Possible keys are "algorithm",
"divisor", "multiplier", "name" and "options".

See L<https://learn.netdata.cloud/docs/agent/collectors/plugins.d#dimension>
for more details.

=head2 to_string

  $str = $chart->to_string;

Creates a string with CHART and DIMENSION, suitable to be sent to Netdata.

=head1 SEE ALSO

L<Mojo::Netdata>.

=cut


lib/Mojo/Netdata/Collector.pm
package Mojo::Netdata::Collector;
use Mojo::Base 'Mojo::EventEmitter', -signatures;

use Carp qw(croak);
use Mojo::Netdata::Chart;
use Mojo::Netdata::Util qw(logf);
use Mojo::Promise;
use Time::HiRes qw(time);

has charts       => sub ($self) { +{} };
has module       => sub ($self) { lc(ref $self) =~ s!\W+!_!gr };
has type         => sub ($self) { croak '"type" cannot be built' };
has update_every => 1;

sub chart ($self, $id) {
  return $self->charts->{$id} //= Mojo::Netdata::Chart->new(
    module       => $self->module,
    id           => $id,
    type         => $self->type,
    update_every => $self->update_every,
  );
}

sub recurring_update_p ($self) {
  my $next_time = time + $self->update_every;

  return $self->{recurring_update_p} //= $self->update_p->then(sub {
    $self->emit_data;
    logf(debug => 'Will update in %0.3fs...', $next_time - time);
    return Mojo::Promise->timer($next_time - time);
  })->then(sub {
    delete $self->{recurring_update_p};
    return $self->recurring_update_p;
  });
}

sub register ($self, $config, $netdata) { }
sub update_p ($self)                    { Mojo::Promise->resolve }

sub emit_data ($self) {
  my @stdout = map { $self->charts->{$_}->data_to_string } sort keys %{$self->charts};
  return $self->emit(stdout => join '', @stdout);
}

sub emit_charts ($self) {
  my @stdout = map { $self->charts->{$_}->to_string } sort keys %{$self->charts};
  return $self->emit(stdout => join '', @stdout);
}

1;

=encoding utf8

=head1 NAME

Mojo::Netdata::Collector - Base class for Mojo::Netdata collectors

=head1 SYNOPSIS

  package Mojo::Netdata::Collector::CoolBeans;
  use Mojo::Base 'Mojo::Netdata::Collector', -signatures;

  has type    => 'ice_cool';

  sub register ($self, $config, $netdata) { ... }
  sub update_p ($self) { ... }

  1;

=head1 DESCRIPTION

L<Mojo::Netdata::Collector> has basic functionality which should be inherited
by L<Mojo::Netdata> collectors. See L<Mojo::Netdata::Collector::HTTP> for an
(example) implementation.

=head1 ATTRIBUTES

=head2 charts

  $hash_ref = $collector->charts;

=head2 module

  $str = $collector->module;

Defaults to a decamelized version of L<$collector>. This is used as default
value for L<Mojo::Netdata::Chart/module>.

=head2 type

  $str = $collector->type;

This value must be set. This is used as default value for
L<Mojo::Netdata::Chart/type>.

=head2 update_every

  $num = $chart->update_every;

Used by L</recurring_update_p> to figure out how often to update Netdata.

=head1 METHODS

=head2 chart

  $chart = $collector->chart($id);

Returns a L<Mojo::Netdata::Chart> object identified by L</id>.

=head2 emit_charts

  $collector = $collector->emit_charts;

Emits all the L</charts> specifications as an "stdout" event.

=head2 emit_data

  $collector = $collector->emit_data;

Emits all the L</charts> data as an "stdout" event.

=head2 recurring_update_p

  $p = $collector->recurring_update_p;

Calls L</update_p> on L</update_every> interval, until the process is killed.

=head2 register

  $collector = $collector->register(\%config, $netdata);

Called by L<Mojo::Netdata> when initialized, and should return C<$collector> on
success or C<undef> if the collector should I<not> be registered.

=head2 update_p

  $p = $collector->update_p;

Must be defined in the sub class. This method should update
L<Mojo::Netdata::Chart/dimensions>.

=head1 SEE ALSO

L<Mojo::Netdata>.

=cut


lib/Mojo/Netdata/Collector/HTTP.pm
package Mojo::Netdata::Collector::HTTP;
use Mojo::Base 'Mojo::Netdata::Collector', -signatures;

use Mojo::UserAgent;
use Mojo::Netdata::Util qw(logf safe_id);
use Time::HiRes         qw(time);

require Mojo::Netdata;
our $VERSION = $Mojo::Netdata::VERSION;

has concurrency => 4;
has jobs        => sub ($self) { +[] };
has type        => 'HTTP';
has ua => sub { Mojo::UserAgent->new(insecure => 0, connect_timeout => 5, request_timeout => 5) };
has update_every => 30;

sub register ($self, $config, $netdata) {
      $config->{update_every}      ? $self->update_every($config->{update_every})
    : $netdata->update_every >= 10 ? $self->update_every($netdata->update_every)
    :                                $self->update_every(30);

  $self->ua->insecure($config->{insecure})               if defined $config->{insecure};
  $self->ua->connect_timeout($config->{connect_timeout}) if defined $config->{connect_timeout};
  $self->ua->request_timeout($config->{request_timeout}) if defined $config->{request_timeout};
  $self->ua->proxy->detect                               if $config->{proxy} // 1;
  $self->ua->transactor->name($config->{user_agent} || "Mojo-Netdata/$VERSION (Perl)");
  $self->concurrency($config->{concurrency} || 4);
  $self->jobs([]);

  my @jobs = ref $config->{jobs} eq 'HASH' ? %{$config->{jobs}} : @{$config->{jobs}};
  while (my $url = shift @jobs) {
    my $job = $self->_make_job($url => ref $jobs[0] eq 'HASH' ? shift @jobs : {}, $config);
    push @{$self->jobs}, $job if $job;
  }

  return @{$self->jobs} ? $self : undef;
}

sub update_p ($self) {
  my ($ua, @p) = ($self->ua);

  return Mojo::Promise->map(
    {concurrency => $self->concurrency},
    sub {
      my ($job, $t0) = ($_, time);
      my $tx = $ua->build_tx(@{$job->[0]});
      return $ua->start_p($tx)->then(sub ($tx) {
        $job->[1]->($tx, $t0);
      })->catch(sub ($err) {
        $job->[1]->($tx, $t0, {message => $err});
      });
    },
    @{$self->jobs}
  );
}

sub _make_job ($self, $url, $params, $defaults) {
  $url = Mojo::URL->new($url);
  return undef unless my $host = $url->host;

  my $headers = Mojo::Headers->new->from_hash($defaults->{headers} || {});
  $headers->header($_ => $params->{headers}{$_}) for keys %{$params->{headers} || {}};
  ($headers->header(Host => $url->host), $url->host($params->{via})) if $params->{via};

  my $dimension = $params->{dimension}   || $headers->host || $url->host;
  my $family    = $params->{family}      || $defaults->{family} || $headers->host || $url->host;
  my $log_level = $defaults->{log_level} || 'debug';

  my $code_chart = $self->chart("${family}_code")->title("HTTP Status code for $family")
    ->context('httpcheck.code')->family($family)->units('#');

  if ($code_chart->dimension($dimension)) {
    logf(warnings => 'Family "%s" already has dimension "%s".', $family, $dimension);
    return undef;
  }

  my $time_chart = $self->chart("${family}_time")->title("Response time for $family")
    ->context('httpcheck.responsetime')->family($family)->units('ms');

  $code_chart->dimension($dimension => {});
  $time_chart->dimension($dimension => {});

  my $update = sub ($tx, $t0, $err = undef) {
    $err ||= $tx->error;
    my $req  = $tx->req;
    my $code = $tx->res->code // 0;
    my @msg  = ($req->method, $req->url, $err || {code => $code}, $req->headers->to_hash(1));
    logf(($code >= 200 && $code < 300 ? $log_level : 'warnings'), '%s %s == %s %s', @msg);

    $time_chart->dimension($dimension => {value => int(1000 * (time - $t0))});
    $code_chart->dimension($dimension => {value => $code});
  };

  my @data;
  push @data, $headers->to_hash(1);
  push @data,
      exists $params->{json} ? (json => $params->{json})
    : exists $params->{form} ? (form => $params->{form})
    : exists $params->{body} ? ($params->{body})
    :                          ();

  my $http_method = $params->{method} || 'GET';
  logf(debug => 'Tracking %s %s %s', $http_method, $url, $data[0]);
  return [[$http_method, $url->to_unsafe_string, @data], $update];
}

1;

=encoding utf8

=head1 NAME

Mojo::Netdata::Collector::HTTP - A website collector for Mojo::Netdata

=head1 SYNOPSIS

=head2 Config

Below is an example C</etc/netdata/mojo.conf.d/http.conf.pl> config file. Note
that the file can have any name and you have have as many as you want, as long
as it has the C<.conf.pl> extension.

  {
    # Required
    collector => 'Mojo::Netdata::Collector::HTTP',

    # Optional
    concurrency     => 4,     # Number of HTTP requests at the same time
    insecure        => 0,     # Set to "1" to allow insecure SSL/TLS connections
    connect_timeout => 5,     # Max time for the connection to be established
    request_timeout => 5,     # Max time for the whole request to complete
    proxy           => 1,     # Set to "0" to disable proxy auto-detect
    update_every    => 30,    # How often to run the "jobs" below
    user_agent      => '...', # Custom User-Agent name

    # Default values, unless defined in the job
    family  => 'default-family-name',
    headers => {'X-Merged-With' => 'headers inside job config'},

    # Required - List of URLs and an optional config hash (object)
    jobs => [

      # List of URLs to check (Config is optional)
      'https://superwoman.example.com',
      'https://superman.example.com',

      # URL and config parameters
      'https://example.com' => {
        method  => 'GET',              # GET (Default), HEAD, POST, ...
        headers => {'X-Foo' => 'bar'}, # HTTP headers

        # Replace "host" in the URL with this IP and set the "Host" header
        via => '192.168.2.1',

        # Set "dimension" to get a custom label in the chart.
        # Default to the "Host" header or the host part of the URL.
        dimension => 'foo', # Default: "example.com"

        # Set "family" to group multiple domains together in one chart,
        # Default to the "Host" header or the host part of the URL.
        family => 'bar', # Default: "example.com"

        # Only one of these can be present
        json   => {...},           # JSON HTTP body
        form   => {key => $value}, # Form data
        body   => '...',           # Raw HTTP body
      },
    ],
  };

=head2 Health

Here is an example C</etc/netdata/health.d/mojo-http.conf> file:

   template: web_server_code
         on: httpcheck.code
      class: Errors
       type: Web Server
  component: HTTP endpoint
     plugin: mojo
     lookup: max -5m absolute foreach *
      every: 1m
       warn: $this >= 300 && $this < 500
       crit: $this >= 500 && $this != 503
         to: webmaster

   template: web_server_up
         on: httpcheck.code
      class: Errors
       type: Web Server
  component: HTTP endpoint
     plugin: mojo
     lookup: min -5m absolute foreach *
      every: 1m
       crit: $this == 0
      units: up/down
         to: webmaster

=head1 DESCRIPTION

L<Mojo::Netdata::Collector::HTTP> is a collector that can chart web page
response time and HTTP status codes.

=head1 ATTRIBUTES

=head2 concurrency

  $int = $collector->concurrency;

Number of requests that should be performed at the same time. Default is 4.

=head2 jobs

  $array_ref = $collector->jobs;

A list of jobs generated by L</register>.

=head2 type

  $str = $collector->type;

Defaults to "HTTP".

=head2 ua

  $ua = $collector->ua;

Holds a L<Mojo::UserAgent>.

=head2 update_every

  $num = $chart->update_every;

Default value is 30. See L<Mojo::Netdata::Collector/update_every> for more
details.

=head1 METHODS

=head2 register

  $collector = $collector->register(\%config, $netdata);

Returns a L<$collector> object if any "jobs" are defined in C<%config>. Will
also set L</update_every> from C<%config> or use L<Mojo::Netdata/update_every>
if it is 10 or greater.

=head2 update_p

  $p = $collector->update_p;

Gathers information about the "jobs" registered.

=head1 SEE ALSO

L<Mojo::Netdata> and L<Mojo::Netdata::Collector>.

=cut


lib/Mojo/Netdata.pm
package Mojo::Netdata;
use Mojo::Base -base, -signatures;

use IO::Handle;
use Mojo::File          qw(path);
use Mojo::Netdata::Util qw(logf);

our $VERSION = '0.04';

has collectors       => sub ($self) { $self->_build_collectors };
has config           => sub ($self) { $self->_build_config };
has user_config_dir  => sub ($self) { $ENV{NETDATA_USER_CONFIG_DIR}  || '/etc/netdata' };
has stock_config_dir => sub ($self) { $ENV{NETDATA_STOCK_CONFIG_DIR} || '/usr/lib/netdata/conf.d' };
has plugins_dir      => sub ($self) { $ENV{NETDATA_PLUGINS_DIR}      || '' };
has web_dir          => sub ($self) { $ENV{NETDATA_WEB_DIR}          || '' };
has cache_dir        => sub ($self) { $ENV{NETDATA_CACHE_DIR}        || '' };
has log_dir          => sub ($self) { $ENV{NETDATA_LOG_DIR}          || '' };
has host_prefix      => sub ($self) { $ENV{NETDATA_HOST_PREFIX}      || '' };
has debug_flags      => sub ($self) { $ENV{NETDATA_DEBUG_FLAGS}      || '' };
has update_every     => sub ($self) { $ENV{NETDATA_UPDATE_EVERY}     || 1 };

sub start ($self) {
  logf(
    info => 'Starting %s with debug_flags=%s host_prefix=%s update_every=%s',
    ref($self), map { $self->$_ } qw(debug_flags host_prefix update_every),
  );
  return 0 unless @{$self->collectors};
  $_->emit_charts->recurring_update_p for @{$self->collectors};
  return int @{$self->collectors};
}

sub _build_config ($self) {
  my $config = {};

  my $file = path($self->user_config_dir, 'mojo.conf.pl')->to_abs;
  if (-r $file) {
    logf(debug => 'Loading config file %s into config().', $file);
    $config = _eval_file($file);
  }

  my $conf_d = path($self->user_config_dir, 'mojo.conf.d')->to_abs->list->sort;
  for my $file ($conf_d->each) {
    next unless $file->basename =~ m!\.conf\.pl$!;
    next unless my $d = _eval_file($file);

    if ($d->{collector}) {
      logf(debug => 'Adding config file %s to "collectors".', $file);
      push @{$config->{collectors}}, $d;
    }
    else {
      logf(debug => 'Merging config file %s into config().', $file);
      @$config{keys(%$d)} = values %$d;
    }
  }

  return $config;
}

sub _build_collectors ($self) {
  my $fh = $self->{stdout} // \*STDOUT;    # for testing
  $fh->autoflush;

  local $@;
  my @collectors;
  for my $collector_config (@{$self->config->{collectors} || []}) {
    my $collector_class = $collector_config->{collector};

    unless ($collector_class and $collector_class =~ m!^[\w:]+$!) {
      logf(error => 'Invalid collector_class %s', $collector_class || 'missing');
      next;
    }
    unless (eval "require $collector_class;1") {
      logf(error => 'Load %s FAIL %s', $collector_class, $@);
      next;
    }

    next unless my $collector = $collector_class->new->register($collector_config, $self);
    $collector->on(stdout => sub ($collector, $str) { $fh->print($str) });
    logf(debug => 'Loaded and set up %s', $collector_class);
    push @collectors, $collector;
  }

  return \@collectors;
}

sub _eval_file ($file) {
  local $@;
  my $prefix = 'package Mojo::Netdata::Config; no warnings; use Mojo::Base -strict;';
  my $config = eval $prefix . $file->slurp;
  logf(error => 'Config file "%s" is invalid: %s', $file, $@) if $@;
  return $config;
}

1;

=encoding utf8

=head1 NAME

Mojo::Netdata - https://netdata.cloud plugin for Perl

=head1 SYNOPSIS

=head2 Installation

  sudo -i
  apt install -y cpanminus
  cpanm -n https://github.com/jhthorsen/mojo-netdata/archive/refs/heads/main.tar.gz
  ln -s $(which mojo-netdata) /etc/netdata/custom-plugins.d/mojo-netdata.plugin

  # See "Config file" below for information on what to place inside mojo.conf.pl
  $EDITOR /etc/netdata/mojo.conf.pl

=head2 Config files

The config files are located in C</etc/netdata/mojo.conf.d>. The files are
plain Perl files, which means you can define variables and call functions. The
only important part is that the last statement in the file is a hash-ref.

Any hash-refs that has the "collector" key will be placed into L</collectors>,
while everything else will be merged into L</config>. Example:

  # /etc/netdata/mojo.conf.d/foo.conf.pl
  {foo => 42, bar => 100}

  # /etc/netdata/mojo.conf.d/bar.conf.pl
  {collector => 'Mojo::Netdata::Collector::HTTP', jobs => []}

The two config files above will result in this L</config>:

  {
    foo => 42,
    bar => 100,
    collectors => [
      {collector => 'Mojo::Netdata::Collector::HTTP', jobs => []},
    },
  }

See L<Mojo::Netdata::Collector::HTTP/SYNOPSIS> for an example config file.

=head2 Log file

The output from this Netdata plugin can be found in
C</var/log/netdata/error.log>.

=head1 DESCRIPTION

L<Mojo::Netdata> is a plugin for L<Netdata|https://netdata.cloud>. It can load
custom L<Mojo::Netdata::Collector> classes and write data back to Netdata on a
given interval.

This module is currently EXPERIMENTAL, and the API might change without
warning.

=head1 ATTRIBUTES

=head2 cache_dir

  $path = $netdata->cache_dir;

Holds the C<NETDATA_CACHE_DIR> environment variable. See
L<https://learn.netdata.cloud/docs/agent/collectors/plugins.d#environment-variables>
for more details.

=head2 config

  $hash_ref = $netdata->config;

Holds the config for L<Mojo::Netdata> and all L</collectors>.

=head2 collectors

  $array_ref = $netdata->collectors;

An array-ref of L<Mojo::Netdata::Collector> objects.

=head2 debug_flags

  $str = $netdata->debug_flags;

Defaults to the C<NETDATA_DEBUG_FLAGS> environment variable. See
L<https://learn.netdata.cloud/docs/agent/collectors/plugins.d#environment-variables>
for more details.

=head2 host_prefix

  $str = $netdata->host_prefix;

Defaults to the C<NETDATA_HOST_PREFIX> environment variable. See
L<https://learn.netdata.cloud/docs/agent/collectors/plugins.d#environment-variables>
for more details.

=head2 log_dir

  $path = $netdata->log_dir;

Holds the C<NETDATA_LOG_DIR> environment variable. See
L<https://learn.netdata.cloud/docs/agent/collectors/plugins.d#environment-variables>
for more details.

=head2 plugins_dir

  $path = $netdata->plugins_dir;

Holds the C<NETDATA_PLUGINS_DIR> environment variable. See
L<https://learn.netdata.cloud/docs/agent/collectors/plugins.d#environment-variables>
for more details.

=head2 stock_config_dir

  $path = $netdata->stock_config_dir;

Holds the C<NETDATA_STOCK_CONFIG_DIR> environment variable. See
L<https://learn.netdata.cloud/docs/agent/collectors/plugins.d#environment-variables>
for more details.

=head2 update_every

  $num = $netdata->update_every;

Defaults to the C<NETDATA_UPDATE_EVERY> environment variable. See
L<https://learn.netdata.cloud/docs/agent/collectors/plugins.d#environment-variables>
for more details.

=head2 user_config_dir

  $path = $netdata->user_config_dir;

Holds the C<NETDATA_USER_CONFIG_DIR> environment variable. See
L<https://learn.netdata.cloud/docs/agent/collectors/plugins.d#environment-variables>
for more details.

=head2 web_dir

  $path = $netdata->web_dir;

Holds the C<NETDATA_WEB_DIR> environment variable. See
L<https://learn.netdata.cloud/docs/agent/collectors/plugins.d#environment-variables>
for more details.

=head1 METHODS

=head2 start

  $bool = $netdata->start;

Reads the L</config> and return 1 if any L</collectors> got registered.

=head1 AUTHOR

Jan Henning Thorsen

=head1 COPYRIGHT AND LICENSE

Copyright (C) Jan Henning Thorsen

This program is free software, you can redistribute it and/or modify it under
the terms of the Artistic License version 2.0.

=head1 SEE ALSO

L<https://learn.netdata.cloud/docs/agent/collectors/plugins.d>.

=cut
